enhancements/operator-framework-api/catalogue-check.md

```diff
@@ -0,0 +1,180 @@
+
+#### Criteria
+
+| Check | Type | Description |
+| ------ | ----- | ----- |
+|To ensure that the operator bundle specified the OCP versions which the project should be supported.|warning|Check if `maxOpenShiftVersion` in CSV **OR** the `com.redhat.openshift.versions` in the Dockerfile is used.
+|To ensure that the maxOpenShiftVersion` syntax value is valid|error|Check if `maxOpenShiftVersion` in CSV is valid
+|To ensure that the index image label `com.redhat.openshift.versions` is valid|error|Check if `com.redhat.openshift.versions` in Dockerfile is a valid range/syntax.
+|To ensure that the index image label `com.redhat.openshift.versions` and `maxOpenShiftVersion` are compatible.|error|Check both values and ensure that they are a valid range. E.g if the `maxOpenShiftVersion` is = `4.5` and the label range is `LABEL com.redhat.openshift.versions:v4.5-v4.7` it is invalid.
+|To ensure that the `minKubeVersion` is compatible with the image label `com.redhat.openshift.versions` and/or the `maxOpenShiftVersion`.|error|Check if `maxOpenShiftVersion` in CSV **AND/OR** the `com.redhat.openshift.versions` in the Dockerfile is used. Then, if `minKubeVersion` was also informed check if the criteria matches. E.g if `minKubeVersion` == `1.22` and the `maxOpenShiftVersion` is == `4.5` then it is invalid.
+
+For the future OCP version which will be using Kubernetes API `1.22`
+| Check | Type | Description |
+| ------ | ----- | ----- |
+|To check if the operator bundle is using the CRD API version which is no longer supported|error| Check if `maxOpenShiftVersion` in CSV **OR** the `com.redhat.openshift.versions` in the Dockerfile is used and has the OCP version which is using K8S API >= `1.22` then, check if the CRD manifests in the bundle are using `apiVersion: apiextensions.k8s.io/v1beta1`. (See [here](https://github.com/operator-framework/operator-sdk/blob/v1.4.2/testdata/go/v2/memcached-operator/bundle/manifests/cache.example.com_memcacheds.yaml#L1) this manifest in the SDK testdata go sample using `go/v2`). Note that in this cases users should be informed that the `apiextensions/v1beta1` was deprecated in Kubernetes `1.16` and was removed in `1.22` and the respective OCP versions as well.
+|To check if the operator bundle is using the Webhook API version which is no longer supported|error|Check if `maxOpenShiftVersion` in CSV **OR** the `com.redhat.openshift.versions` in the Dockerfile is used and has the OCP version which is using K8S API >= `1.22` then, check if the CSV has the spec `webhookdefinitions.admissionReviewVersions.v1beta1` . (See [here](https://github.com/operator-framework/operator-sdk/blob/v1.4.2/testdata/go/v2/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml#L203-L205). Note that in this cases users should be informed that the `admissionregistration.k8s.io/v1beta1` was deprecated in Kubernetes `1.16` and is removed in `1.22` and the respective OCP versions as well.
```

Comment:
I had a slightly different model in mind.  Rather than internalizing the logic that says "which versions does this bundle claim to support, now validate the bundle against rules for that version" I was anticipating something more decoupled in which the user specifies which versions(rules) they want to validate against.  

There could reasonably be an auto/auto-detect mode that chooses which version/rules based on the bundle metadata, but i think it's useful to allow any bundle to be validated against any ocp version's rules and see what errors/warnings come out.

Reply:
I understand your motivation and I agree that it is a very reasonable requirement too. However, I think we need to address 2 scenarios:
- I want to check my bundle according to its configuration regards the versions supported
- I want to check my bundle according to its configuration regards against a specific OCP version that I intend to publish. 

I will try to address that here. Really thank you for your support and suggestion.